sw/device/silicon_creator/mask_rom/bootstrap.c

```diff
@@ -0,0 +1,321 @@
+ * @return Result of the operation.
+ */
+static rom_error_t bootstrap_sector_erase(uint32_t addr) {
+  static_assert(__builtin_popcount(FLASH_CTRL_PARAM_BYTES_PER_PAGE) == 1,
+                "Bytes per page must be a power of two.");
+  enum {
+    /**
+     * Mask for checking that `addr` is page aligned.
+     */
+    kPageAddrMask = FLASH_CTRL_PARAM_BYTES_PER_PAGE - 1,
+  };
+
+  if (addr & kPageAddrMask || addr >= kMaxAddress) {
+    return kErrorBootstrapEraseAddress;
+  }
+
+  flash_ctrl_data_default_perms_set((flash_ctrl_perms_t){
+      .read = kHardenedBoolFalse,
+      .write = kHardenedBoolFalse,
+      .erase = kHardenedBoolTrue,
+  });
+  rom_error_t err_0 = flash_ctrl_data_erase(addr, kFlashCtrlEraseTypePage);
+  rom_error_t err_1 = flash_ctrl_data_erase(
+      addr + FLASH_CTRL_PARAM_BYTES_PER_PAGE, kFlashCtrlEraseTypePage);
+  flash_ctrl_data_default_perms_set((flash_ctrl_perms_t){
+      .read = kHardenedBoolFalse,
+      .write = kHardenedBoolFalse,
+      .erase = kHardenedBoolFalse,
+  });
+
+  HARDENED_RETURN_IF_ERROR(err_0);
+  return err_1;
+}
+
+/**
+ * Handles access permissions and programs up to 256 bytes of flash memory
+ * starting at `addr`.
+ *
+ * If `byte_count` is not a multiple of flash word size, it's rounded up to next
+ * flash word and missing bytes in `data` are set to `0xff`.
+ *
+ * @param addr Address to write to, must be flash word aligned.
+ * @param byte_count Number of bytes to write. Rounded up to next flash word if
+ * not a multiple of flash word size. Missing bytes in `data` are set to `0xff`.
+ * @param data Data to write, must be word aligned. If `byte_count` is not a
+ * multiple of flash word size, `data` must have enough space until the next
+ * flash word.
+ * @return Result of the operation.
+ */
+static rom_error_t bootstrap_page_program(uint32_t addr, size_t byte_count,
+                                          uint8_t *data) {
+  static_assert(__builtin_popcount(FLASH_CTRL_PARAM_BYTES_PER_WORD) == 1,
+                "Bytes per flash word must be a power of two.");
+  enum {
+    /**
+     * Mask for checking that `addr` is flash word aligned.
+     */
+    kWordAddrMask = FLASH_CTRL_PARAM_BYTES_PER_WORD - 1,
+  };
+
+  if (addr & kWordAddrMask || addr >= kMaxAddress) {
+    return kErrorBootstrapProgramAddress;
+  }
+
+  // Round up to next flash word and fill missing bytes with `0xff`.
+  size_t tail_byte_count = byte_count & kWordAddrMask;
+  if (tail_byte_count > 0) {
+    size_t padding_byte_count =
+        FLASH_CTRL_PARAM_BYTES_PER_WORD - tail_byte_count;
+    for (size_t i = 0; i < padding_byte_count; ++i) {
+      data[byte_count++] = 0xff;
+    }
+  }
+  size_t word_count = byte_count / sizeof(uint32_t);
+
+  flash_ctrl_data_default_perms_set((flash_ctrl_perms_t){
+      .read = kHardenedBoolFalse,
+      .write = kHardenedBoolTrue,
+      .erase = kHardenedBoolFalse,
+  });
+  rom_error_t error = flash_ctrl_data_write(addr, word_count, data);
```

Comment:
The flash programming page size is 256B.

What will happen if I ask to program at address 0xF0 and I send 256 bytes?
- Program from 0x0F0 to 0x1F0?
- Program from 0xF0 to 0xFF, then wrap to 0x00 and program up to 0xF0.

I believe a real SPI eeprom will perform the latter.

Reply:
I checked this behavior on an MX25L512 part.  That EEPROM wraps around within the page referred to by the start address.

(I used opentitantool, but I had to disable my own error checking in //sw/host/opentitanlib/src/spiflash/flash.rs)

Reply:
Hum.. if the behavior is consistent, I think upload function also should behave same rather than uploading from `0x00` offset in Payload buffer always.. One more TODO..

Reply:
@eunchan I'm not sure if you plan to handle this in HW, but this is something that we can handle in SW.

Reply:
If it is easy to handle in SW, please do so. Thanks!

Reply:
Done, this part should now work like @cfrantz described.